File: DisasterServer/Entities/PFLift.cs
using DisasterServer.Maps;
using DisasterServer.Session;
using DisasterServer.State;

namespace DisasterServer.Entities;

internal class PFLift : Entity
{
	public byte ID;

	public ushort ActivatorID;

	private bool _activated;

	private float _y;

	private float _startY;

	private float _endY;

	private int _timer;

	private float _speed;

	public PFLift(byte id, float starty, float endY)
	{
		ID = id;
		_startY = starty;
		_endY = endY;
	}

	public override TcpPacket? Destroy(Server server, Game game, Map map)
	{
		return null;
	}

	public override TcpPacket? Spawn(Server server, Game game, Map map)
	{
		return null;
	}

	public override UdpPacket? Tick(Server server, Game game, Map map)
	{
		if (!_activated)
		{
			if (_timer > 0)
			{
				_timer--;
				if (_timer == 0)
				{
					server.TCPMulticast(new TcpPacket(PacketType.SERVER_PFLIFT_STATE, (byte)3, ID, (ushort)_startY));
				}
			}
			return null;
		}
		if (_y > _endY)
		{
			if (_speed < 7f)
			{
				_speed += 0.052f;
			}
			_y -= _speed;
		}
		else
		{
			server.TCPMulticast(new TcpPacket(PacketType.SERVER_PFLIFT_STATE, (byte)2, ID, ActivatorID, (ushort)_y));
			_timer = 90;
			_activated = false;
			ActivatorID = 0;
		}
		return new UdpPacket(PacketType.SERVER_PFLIFT_STATE, (byte)1, ID, ActivatorID, (ushort)_y);
	}

	public void Activate(Server server, ushort id)
	{
		if (!_activated && _timer <= 0)
		{
			ActivatorID = id;
			_timer = 0;
			_speed = 0f;
			_y = _startY;
			_activated = true;
			server.TCPMulticast(new TcpPacket(PacketType.SERVER_PFLIFT_STATE, (byte)0, ID, ActivatorID));
		}
	}
}

File: disasterserver/entities/PFLift.c
#include <entities/PFLift.h>

bool pflift_init(Server* server, Entity* entity)
{
	(void)server;
	
	PFLift* lift = (PFLift*)entity;
	lift->pos.y = lift->start;

	return true;
}

bool pflift_tick(Server* server, Entity* entity)
{
	Packet pack;
	PFLift* lift = (PFLift*)entity;

	if (!lift->activated)
	{
		if (lift->timer > 0)
		{
			lift->timer -= server->delta;
			if (lift->timer <= 0)
			{
				// teleport back
				PacketCreate(&pack, SERVER_PFLIFT_STATE);
				PacketWrite(&pack, packet_write8, 3);
				PacketWrite(&pack, packet_write8, lift->lid);
				PacketWrite(&pack, packet_write16, (uint16_t)lift->start);
				server_broadcast(server, &pack, true);
			}
		}
		return true;
	}

	if (lift->pos.y > lift->end)
	{
		if (lift->speed < 7.f)
			lift->speed += 0.052f * (float)server->delta;

		lift->pos.y -= lift->speed * (float)server->delta;
	}
	else
	{
		PacketCreate(&pack, SERVER_PFLIFT_STATE);
		PacketWrite(&pack, packet_write8, 2);
		PacketWrite(&pack, packet_write8, lift->lid);
		PacketWrite(&pack, packet_write16, lift->activator);
		PacketWrite(&pack, packet_write16, (uint16_t)lift->pos.y);
		server_broadcast(server, &pack, true);
		
		lift->timer = (uint16_t)(1.5f * TICKSPERSEC);
		lift->activated = false;
		lift->activator = 0;
	}

	PacketCreate(&pack, SERVER_PFLIFT_STATE);
	PacketWrite(&pack, packet_write8, 1);
	PacketWrite(&pack, packet_write8, lift->lid);
	PacketWrite(&pack, packet_write16, lift->activator);
	PacketWrite(&pack, packet_write16, (uint16_t)lift->pos.y);
	server_broadcast(server, &pack, false);

	return true;
}

bool pflift_activate(Server* server, PFLift* lift, uint16_t id)
{
	if (lift->activated)
		return true;

	if (lift->timer > 0)
		return true;

	lift->activator = id;
	lift->timer = 0;
	lift->speed = 0;
	lift->pos.y = lift->start;
	lift->activated = true;

	Packet pack;
	PacketCreate(&pack, SERVER_PFLIFT_STATE);
	PacketWrite(&pack, packet_write8, 0);
	PacketWrite(&pack, packet_write8, lift->lid);
	PacketWrite(&pack, packet_write16, lift->activator);
	server_broadcast(server, &pack, true);

	return true;
}
